fedimint-client/src/meta.rs

```diff
@@ -0,0 +1,371 @@
+        field: &str,
+    ) -> Option<TypedMetaValue<V>> {
+        if let Some(value) = self.get_field_inner(db, field).await {
+            // might be from in old cache.
+            // TODO: maybe old cache should have a ttl?
+            Some(value)
+        } else {
+            // wait for initial value
+            self.initial_fetch_waiter.wait().await;
+            self.get_field_inner(db, field).await
+        }
+    }
+
+    async fn get_field_inner<V: DeserializeOwned + 'static>(
+        &self,
+        db: &Database,
+        field: &str,
+    ) -> Option<TypedMetaValue<V>> {
+        db.begin_transaction_nc()
+            .await
+            // sorted by priority
+            .find_by_prefix_sorted_descending(&MetaFieldName(field.to_string()))
+            .await
+            .collect::<Vec<_>>()
+            .await
+            .into_iter()
+            .find_map(|(key, value)| {
+                Some(TypedMetaValue {
+                    revision: value.revision,
+                    value: parse_meta_value_static::<V>(&value.value).ok()?,
+                    priority: key.priority,
+                })
+            })
+    }
+
+    pub fn add_source(&mut self, source: impl MetaSource + 'static) {
+        self.sources.push(Box::new(source))
+    }
+
+    /// Update all source in background.
+    ///
+    /// Caller should run this method in a task.
+    pub async fn update_continuously(&self, client: &Client) -> ! {
+        // need to be fast the first time.
+        let metas = join_all(self.sources.iter().map(|source| async move {
+            (
+                source,
+                timeout(
+                    Duration::from_secs(60),
+                    retry(
+                        "update_source",
+                        FibonacciBackoff::default()
+                            .with_min_delay(Duration::from_millis(300))
+                            .with_max_delay(Duration::from_secs(3)),
+                        || source.fetch(client),
+                    ),
+                )
+                .await
+                .context("timeout while fetch source"),
+            )
+        }))
+        .await;
+        // initially write all once
+        let mut dbtx = client.db().begin_transaction().await;
+        for (source, meta) in metas {
+            if let Ok(meta) = meta.and_then(|x| x) {
+                Self::update_meta_values(&mut dbtx, source.priority(), source.source_id(), &meta)
+                    .await;
+            }
+        }
+        dbtx.commit_tx().await;
+        self.initial_fetch_waiter.done();
+
+        // now keep updating all
+        let values = join_all(
+            self.sources
+                .iter()
+                .map(|source| self.update_source_continuously(client, &**source)),
+        )
+        .await;
+        // to get ! type
+        values[0]
+    }
+
+    async fn update_source_continuously(&self, client: &Client, source: &dyn MetaSource) -> ! {
```

Comment:
Since this is *always* started now running it twice doesn't sound too useful, thoughts about making the method private?

Reply:
it is private :thinking: , you mean the other one

Reply:
fixed

Reply:
Oh yeah, but fixed I see :sweat_smile: